### PR TITLE
Fixes compatibility with StatsBase 0.4

### PR DESCRIPTION
**Pull request by racinmat**
Changes needed in order to have #45 working. 
Locally it works, tried with MLLabelUtils fork https://github.com/JuliaML/MLLabelUtils.jl/pull/40.
I hope it's correct, all tests are passing.

### CODE REVIEW

**Comment by racinmat** [comment]
@johnnychen94

File: src/MLDataPattern.jl [hunk 1]
```diff
@@ -4,7 +4,9 @@ using LearnBase
 using MLLabelUtils
 
 using LearnBase: ObsDimension
-import LearnBase: nobs, getobs, getobs!, gettarget, gettargets, targets, datasubset, default_obsdim
+import StatsBase: nobs
+import LearnBase: getobs, getobs!, gettarget, gettargets, targets, datasubset, default_obsdim, DataView,
```
Comment (johnnychen94) [inline src/MLDataPattern.jl:8]: Are these unexported by LearnBase 0.4?

`DataView, AbstractObsView, AbstractBatchView, DataIterator, AbstractDataIterator, ObsIterator, BatchIterator`

Comment (racinmat) [reply src/MLDataPattern.jl:8]: AFAIK yes
```julia
julia> using LearnBase
julia> DataView
ERROR: UndefVarError: DataView not defined

julia> AbstractObsView
ERROR: UndefVarError: AbstractObsView not defined

julia> AbstractBatchView
ERROR: UndefVarError: AbstractBatchView not defined

julia> DataIterator
ERROR: UndefVarError: DataIterator not defined

julia> AbstractDataIterator
ERROR: UndefVarError: AbstractDataIterator not defined

julia> ObsIterator
ERROR: UndefVarError: ObsIterator not defined

julia> BatchIterator
ERROR: UndefVarError: BatchIterator not defined

julia> LearnBase.BatchIterator
LearnBase.BatchIterator
```

Comment (racinmat) [reply src/MLDataPattern.jl:8]: There was quite a lot of unexporting in https://github.com/JuliaML/LearnBase.jl/compare/v0.3.0...v0.4.0

**Review by johnnychen94 — APPROVED** [review]
Good to merge as long as the test passes

**Comment by johnnychen94** [comment]
But keep in mind that once https://github.com/JuliaML/LearnBase.jl/pull/44 is merged, we will need to handle 0.5 compatibilities here, and that would be way more complex since the whole ObsDim trick is removed.

File: src/MLDataPattern.jl (hunk 1 above)
Comment (darsnack) [reply src/MLDataPattern.jl:8]: I can export these in https://github.com/JuliaML/LearnBase.jl/pull/44 if that is the preferred end result. Seems like a lot got lost in translation between 0.3 and 0.4.

**Comment by racinmat** [comment]
Yes, I'm aware of that, thanks. But hopefully the interface will be simpler that, using dims the cleaner way.

**Comment by darsnack** [comment]
I need to be diligent about making sure everything still works, but on my local copy, I was able to replace all the `ObsDim` tricks and reduce the codebase.

**Comment by racinmat** [comment]
It would be great if you shared the diff, I think some small "migration guide" would be helpful given the fact that multiple libraries will be affected

**Comment by darsnack** [comment]
Absolutely, I just need to do a little clean-up since it's a mess of commenting out parts and deleting parts right now. But I'll share something soon.

**Comment by darsnack** [comment]
[Here is a PR](https://github.com/JuliaML/MLDataPattern.jl/pull/50) that reflects the diff. I also included a summary of my experience.

**Comment by racinmat** [comment]
Hm, there are some problems with dependencies, most probably related to ReferenceTests and its image related functions for Julia 1.0, I have no idea how to fix it. Maybe restrict version of ReferenceTests?

**Comment by darsnack** [comment]
Yeah I am not surprised. I didn't commit Project.toml because I have local copies of LearnBase, MLLabelUtils, and MLDataPattern floating around. I ran into a bunch of issues trying to get those dependencies to work together without version issues or method errors once I deprecated `ObsDim`. I'll try and clean up the dependency mess.

**Comment by johnnychen94** [comment]
> Hm, there are some problems with dependencies, most probably related to ReferenceTests and its image related functions for Julia 1.0, I have no idea how to fix it. Maybe restrict version of ReferenceTests?

Should be fixed after https://github.com/JuliaRegistries/General/pull/23857

**Comment by racinmat** [comment]
I see that in nightly, the string representation of type is different, 
e.g. it returns 
```julia
julia> typeof(zeros(2,3))
Matrix{Float64} (alias for Array{Float64, 2})
```
instead of
```julia
julia> typeof(zeros(2,3))
Array{Float64,2}
```
this should be also handled by custom comparison function?

**Comment by johnnychen94** [comment]
Yes, for example, in https://github.com/JuliaML/MLDataPattern.jl/blob/2cf8d0ec79475207539c2e898dcbb35c245ca61b/test/tst_datasubset.jl#L50

```diff
+ strip_summary(content::String) = join(split(content, "\n")[2:end], "\n")

- @test_reference "references/DataSubset2.txt" @io2str showcompact(::IO, DataSubset(X))
+ @test_reference "references/DataSubset2.txt" stripe_summary(@io2str(showcompact(::IO, DataSubset(X))))
```

Such an example can be found in https://github.com/JuliaTesting/ReferenceTests.jl/blob/c8104a1628295d9e2ce79f58a29a97f0c3dc26bf/test/runtests.jl#L58

**Comment by racinmat** [comment]
I see, thanks. Done.

**Comment by johnnychen94** [comment]
lol the reference files in `test/references` might also need some update.

**Comment by racinmat** [comment]
And what exactly? Because they are passing on Julia 1.0 and 1, right?
In the end I used different comparison function so I don't have to call `show` explicitly in the `@test_reference "references/DataSubset1.txt" DataSubset(X, Int64(1):Int64(nobs(X)))` testcase

**Comment by johnnychen94** [comment]
@racinmat Thanks for updating the codes!

**Pull request merged**